engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/MergedDataIndex.java

```diff
@@ -113,65 +140,185 @@ public Map<ColumnSource<?>, String> keyColumnNamesByIndexedColumn() {
-                        columnSourceManager.locationColumnName(), TableLocation.class,
-                        LOCATION_DATA_INDEX_TABLE_COLUMN_NAME, Table.class,
-                        (final long locationRowKey, final TableLocation location) -> loadIndexTableAndShiftRowSets(
-                                locationRowKey, location, keyColumnNamesArray)))))
-                .dropColumns(columnSourceManager.locationColumnName());
-
-        // Merge all the location index tables into a single table
-        final Table mergedDataIndexes = PartitionedTableFactory.of(locationDataIndexes).merge();
-
-        // Group the merged data indexes by the keys
-        final Table groupedByKeyColumns = mergedDataIndexes.groupBy(keyColumnNamesArray);
-
-        // Combine the row sets from each group into a single row set
-        final Table combined = groupedByKeyColumns
-                .update(List.of(SelectColumn.ofStateless(new FunctionalColumn<>(
+        private RowsetCacher(final ColumnSource<ObjectVector<RowSet>> source, final int capacity) {
+            this.source = source;
+            this.results = new AtomicReferenceArray<>(capacity);
+        }
+
+        RowSet get(final long rowKey) {
+            if (rowKey < 0 || rowKey >= results.length()) {
+                return null;
+            }
+
+            final int iRowKey = (int) rowKey;
+            do {
+                final Object localResult = results.get(iRowKey);
+                if (localResult instanceof RowSet) {
+                    return (RowSet) localResult;
+                }
+                if (localResult instanceof Exception) {
+                    throw new UncheckedDeephavenException("Exception found for cached RowSet",
+                            (Exception) localResult);
+                }
+
+                if (localResult != null) {
+                    // noinspection EmptySynchronizedStatement
+                    synchronized (localResult) {
+                        // don't care to do anything, we are just waiting for the barrier to be done
+                    }
+                    continue;
+                }
+
+                // we need to create our own placeholder, and synchronize on it first
+                final ReentrantLock placeholder = new ReentrantLock();
+                // noinspection SynchronizationOnLocalVariableOrMethodParameter
+                synchronized (placeholder) {
+                    if (!results.compareAndSet(iRowKey, null, placeholder)) {
+                        // we must try again, someone else has claimed the placeholder
+                        continue;
+                    }
+                    // it is our responsibility to get the right answer
+                    final ObjectVector<RowSet> inputRowsets = source.get(rowKey);
+                    Assert.neqNull(inputRowsets, "inputRowsets");
+                    assert inputRowsets != null;
```

Comment:
Maybe just a `noinspection` instead?

Reply:
I put in a noinspection; though I think a better pattern would be to change the Assert.neqNull to return the value with a @NotNull annotation, to solve the data flow problems in a way that IJ can understand.